### ISSUE TITLE
Reuse cache_req() in responder code

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#1130]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/1126
- Created at 2012-01-03 09:25:16 by [jzeleny](https://pagure.io/user/jzeleny)
- Closed at 2017-03-30 13:39:28 as Fixed
- Assigned to [pbrezina](https://pagure.io/user/pbrezina)
---

The function nss_cmd_getpwnam_search() has several slightly modified copies through the code in responders. It would be nice to create one function (or macro) which would be generic and used in all parts of the code instead of those copies.

### Comments

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-01-03 15:20:09](https://pagure.io/SSSD/sssd/issue/1126#225555)**

There are a number of improvements we should work on while fixing this. In particular, we should seek to resolve the performance issue identified in bug #843.

The current approach behaves as follows (I will use {{{getpwnam()}}} as the representative function, though all are equivalent).

The NSS responder will check the cache for the first domain in the list. If it is empty or expired, the Responder will query the Data Provider. If the Data Provider returns DP_ERR_OK, the cache is checked again. If the entry is still not present, it advances to the next domain. This pattern repeats.

The problem with this approach is that it means that for a configuration of N domains, if a user is found in the last domain in the list, SSSD will go to the server N-1 times before ever checking the cache of the correct domain.

The proposed new approach is for us to search each domain's cache first. If any domain has the requested, non-expired entry, it should be returned immediately. If all domains lack the entry, only then should we begin to iterate through them. For NSS lookups, we should NOT limit ourselves only to the domain with an expired entry for the user, because we need to be able to catch new users being added to earlier domains. So in the complete cache-miss case, we will still iterate through all domains, but a cache-hit will not result in network activity.

For PAM, we will take an additional shortcut. Since it is reasonable to assume that the login program will have made a {{{getpwnam()}}} call prior to starting the PAM conversation, we will limit our mandatory {{{initgroups()}}} lookup to only domains with an expired entry (if one exists, otherwise we will try them in series).

blocking:  => 843

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-01-05 15:55:30](https://pagure.io/SSSD/sssd/issue/1126#225556)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.9.0 NEEDS_TRIAGE

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-01-16 16:35:36](https://pagure.io/SSSD/sssd/issue/1126#225557)**

Fields changed

blocking: 843 =>

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-01-16 16:59:59](https://pagure.io/SSSD/sssd/issue/1126#225558)**

"Nice to have" for 1.9.

milestone: SSSD 1.9.0 NEEDS_TRIAGE => SSSD 1.9.0

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-02-09 15:37:19](https://pagure.io/SSSD/sssd/issue/1126#225559)**

Fields changed

feature_milestone:  => 
priority: major => minor

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-02-10 21:38:12](https://pagure.io/SSSD/sssd/issue/1126#225560)**

Fields changed

rhbz:  => 0

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-04-23 18:43:18](https://pagure.io/SSSD/sssd/issue/1126#225561)**

Implementing this sooner will ease work for Pavel. I'm going to squash this one out while he's doing prep work to help him.

owner: somebody => jhrozek

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-05-17 21:04:29](https://pagure.io/SSSD/sssd/issue/1126#225562)**

When implementing this, use the code written for getserv_send(), lookup_service_step() and lookup_service_done() from nssrv_services.c as the guideline. That approach will address Bug #843 as well.

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-08-16 19:14:43](https://pagure.io/SSSD/sssd/issue/1126#225563)**

Fields changed

proposed_priority:  => Nice to have

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-08-16 19:21:38](https://pagure.io/SSSD/sssd/issue/1126#225564)**

Fields changed

milestone: SSSD 1.9.0 => Temp milestone

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-08-16 19:23:37](https://pagure.io/SSSD/sssd/issue/1126#225565)**

Fields changed

blocking:  => 843

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-08-16 19:27:19](https://pagure.io/SSSD/sssd/issue/1126#225566)**

Fields changed

blocking: 843 => #843

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-09-04 23:20:30](https://pagure.io/SSSD/sssd/issue/1126#225567)**

Moving all the features planned for 1.10 release into 1.10 beta.

milestone: Temp milestone => SSSD 1.10 beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2012-12-20 23:33:42](https://pagure.io/SSSD/sssd/issue/1126#225568)**

Fields changed

selected:  => Not need

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2013-01-02 15:13:11](https://pagure.io/SSSD/sssd/issue/1126#225569)**

Fields changed

design:  => 
design_review:  => 0
fedora_test_page:  => 
selected: Not need => Want

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2013-01-03 16:30:11](https://pagure.io/SSSD/sssd/issue/1126#225570)**

Fields changed

selected: Want => Not need

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2013-03-16 01:19:53](https://pagure.io/SSSD/sssd/issue/1126#225571)**

Fields changed

review:  => 1

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2013-03-28 14:07:25](https://pagure.io/SSSD/sssd/issue/1126#225572)**

Fields changed

milestone: SSSD 1.10 beta => SSSD 1.11 beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2014-02-06 17:09:26](https://pagure.io/SSSD/sssd/issue/1126#225573)**

Putting up to NEEDS_TRIAGE so we can discuss the priority on the next meeting.

changelog:  => 
keywords:  => localaccounts
milestone: SSSD 1.13 beta => NEEDS_TRIAGE

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2014-02-13 16:00:34](https://pagure.io/SSSD/sssd/issue/1126#225574)**

Fields changed

milestone: NEEDS_TRIAGE => SSSD 1.12 beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2014-05-30 16:04:39](https://pagure.io/SSSD/sssd/issue/1126#225575)**

Would be nice to have for local users support.

milestone: SSSD 1.12 beta => SSSD 1.13 beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2014-09-26 18:49:56](https://pagure.io/SSSD/sssd/issue/1126#225576)**

Fields changed

mark:  => 0

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2015-02-10 17:22:41](https://pagure.io/SSSD/sssd/issue/1126#225577)**

We should use Pavel's cache request in the NSS responder instead.

keywords: localaccounts => 
priority: minor => critical

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2015-06-24 16:25:13](https://pagure.io/SSSD/sssd/issue/1126#225578)**

Migrating to Pavel's cache_req should be done in 1.14, it's too late for 1.13 now

milestone: SSSD 1.13 beta => SSSD 1.14 beta
sensitive:  => 0

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2015-07-22 13:54:36](https://pagure.io/SSSD/sssd/issue/1126#225579)**

Fields changed

owner: jhrozek => pbrezina

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2015-11-18 09:51:45](https://pagure.io/SSSD/sssd/issue/1126#225580)**

Fields changed

blockedby:  => #2896
summary: Reuse nss_cmd_getpwnam_search() in responder code => Reuse cache_req() in responder code

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2015-11-18 11:04:36](https://pagure.io/SSSD/sssd/issue/1126#225581)**

Fields changed

blockedby: #2896 => #2869

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-02-11 11:35:16](https://pagure.io/SSSD/sssd/issue/1126#225582)**

Some patches are already landing on the list, so I'm moving this ticket to 1.14 alpha.

milestone: SSSD 1.14 beta => SSSD 1.14 alpha

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-04-19 13:22:10](https://pagure.io/SSSD/sssd/issue/1126#225583)**

Sudo related patches:

master:

- b3ca35780617b2e5a7637f9888b089e8e26a4e8c
- 15d41c8f28259061e39715acdbbbaea778b6ecc8
- 52300e30a0ec0bbfa1b0918ee0b495f06f5d142a
- 536dcc7fb975acfc126846a889d90332304e88ba

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-06-20 11:59:06](https://pagure.io/SSSD/sssd/issue/1126#225584)**

sudo is done, but more work in the other responders is needed.

milestone: SSSD 1.14 alpha => SSSD 1.15 beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-08-22 12:48:40](https://pagure.io/SSSD/sssd/issue/1126#225585)**

Fields changed

blockedby: #2869 => #2869, #3151

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-10-05 22:33:31](https://pagure.io/SSSD/sssd/issue/1126#225586)**

Fields changed

milestone: SSSD 1.16 beta => SSSD 1.15 Alpha

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2016-11-25 10:47:37](https://pagure.io/SSSD/sssd/issue/1126#225587)**

Fields changed

patch: 0 => 1

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-01-25 12:15:38](https://pagure.io/SSSD/sssd/issue/1126#225588)**

Fields changed

milestone: SSSD 1.15.0 => SSSD 1.15 Beta

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-02-06 14:19:25](https://pagure.io/SSSD/sssd/issue/1126#225589)**

Fields changed

milestone: SSSD 1.16 Beta => SSSD 1.15.1

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-02-08 11:15:44](https://pagure.io/SSSD/sssd/issue/1126#225590)**

ssh responder related patches:
    - a8191ce7ad5364801ad9458c3194075a7ca77b8a
    - 53c31b83e4d06ea4c2813eec2f1e647a613b4a2b
    - 4df7aec645f87342f3a5146062abcb15f71f4fd9
    - 9492b3b26ac0b1898f836094074a9d8b38916e13
    - 7723e79f5a1fad4201360199037aea33f655bab6
    - ddfd1900b26c66a062457d4fcc1a48bafd3eadf6
    - 2b5704cd96a085b99d3b0d4f80f4414adc134750
    - e33744e8cc82390153c94ace53c16f72365b9fd9
    - d8c459feab7659a51c23c941fea486867c2b9dae

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-02-08 21:28:48](https://pagure.io/SSSD/sssd/issue/1126#225591)**

ssh fixes:

master:

- d9780d2860b2f2c9d707bfd8f2fc72099b9545d7
- 2ffa245e79a5ed66e69d141f4001c13697e01450

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-02-24 14:31:59](https://pagure.io/SSSD/sssd/issue/1126#225592)**

**Metadata Update from @jzeleny**:
- Issue assigned to pbrezina
- Issue marked as blocked by: #2228
- Issue marked as blocked by: #843
- Issue marked as depending on: #2869
- Issue marked as depending on: #3151
- Issue set to the milestone: SSSD 1.15.1

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-03-03 14:00:05](https://pagure.io/SSSD/sssd/issue/1126#429276)**

pam fixes:

master:
c99bcc91e58988ff37403d362d4322469413c2ff
5aaaf081765b3f23e1518b5f299c289afb9d3f13
ed891c0c55985cd25de05f65e82debf4452987e1
189db5329db0277425bb3bdd785c19aee23bb364

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-03-03 14:00:10](https://pagure.io/SSSD/sssd/issue/1126#429277)**

pam fixes:

master:
c99bcc91e58988ff37403d362d4322469413c2ff
5aaaf081765b3f23e1518b5f299c289afb9d3f13
ed891c0c55985cd25de05f65e82debf4452987e1
189db5329db0277425bb3bdd785c19aee23bb364

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-03-30 13:28:21](https://pagure.io/SSSD/sssd/issue/1126#434213)**

**Metadata Update from @jhrozek**:
- Custom field blocking reset (from #843)
- Custom field design_review reset (from 0)
- Custom field mark reset (from 0)
- Custom field patch adjusted to on (was: 1)
- Custom field review adjusted to on (was: 1)
- Custom field sensitive reset (from 0)
- Custom field testsupdated reset (from 0)
- Issue **un**marked as blocking: #2869
- Issue **un**marked as blocking: #3151
- Issue **un**marked as depending on: #2228
- Issue **un**marked as depending on: #843

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-03-30 13:39:01](https://pagure.io/SSSD/sssd/issue/1126#434214)**

**Metadata Update from @jhrozek**:
- Custom field design_review reset (from false)
- Custom field mark reset (from false)
- Custom field sensitive reset (from false)
- Custom field testsupdated reset (from false)

---

**Comment from [jzeleny](https://pagure.io/user/jzeleny) at [2017-03-30 13:39:31](https://pagure.io/SSSD/sssd/issue/1126#434215)**

**Metadata Update from @jhrozek**:
- Custom field design_review reset (from false)
- Custom field mark reset (from false)
- Custom field sensitive reset (from false)
- Custom field testsupdated reset (from false)
- Issue close_status updated to: Fixed
- Issue status updated to: Closed (was: Open)
